File: scripts/verify-stryker-gate.py
# Copyright © Erickson Lopez. MIT License.
import json
import os
import sys
import urllib.request
import urllib.error

if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8", errors="backslashreplace")
if hasattr(sys.stderr, "reconfigure"):
    sys.stderr.reconfigure(encoding="utf-8", errors="backslashreplace")

def load_thresholds(config_path="stryker-config.json"):
    thresholds = {"high": 100, "low": 98, "break": 95}
    if os.path.exists(config_path):
        try:
            with open(config_path, "r", encoding="utf-8") as f:
                config = json.load(f)
            t = config.get("stryker-config", {}).get("thresholds", config.get("thresholds", {}))
            thresholds = {
                "high": t.get("high", 100),
                "low": t.get("low", 98),
                "break": t.get("break", 95)
            }
        except Exception as e:
            print(f"Warning: Could not parse {config_path}: {e}")
    return thresholds

def fetch_json(url, token=None):
    req = urllib.request.Request(url, method="GET")
    req.add_header("User-Agent", "dotnet-idempotency-release-gate")
    req.add_header("Accept", "application/vnd.github.v3+json")
    if token:
        req.add_header("Authorization", f"Bearer {token}")
    try:
        with urllib.request.urlopen(req) as resp:
            if 200 <= resp.status < 300:
                return json.loads(resp.read().decode("utf-8"))
    except Exception as e:
        print(f"Warning: API request error for {url}: {e}")
    return None

def query_commit_status(owner, repo, sha, token):
    url = f"https://api.github.com/repos/{owner}/{repo}/commits/{sha}/statuses"
    statuses = fetch_json(url, token)
    if isinstance(statuses, list):
        for s in statuses:
            if s.get("context") == "quality-gate/stryker-mutation":
                return {
                    "source": "commit_status",
                    "state": s.get("state"),
                    "description": s.get("description"),
                    "target_url": s.get("target_url"),
                    "created_at": s.get("created_at"),
                    "updated_at": s.get("updated_at")
                }
    return None

def query_workflow_runs(owner, repo, sha, token):
    url = f"https://api.github.com/repos/{owner}/{repo}/actions/workflows/mutation-testing.yml/runs?head_sha={sha}&status=completed"
    data = fetch_json(url, token)
    if data and "workflow_runs" in data and len(data["workflow_runs"]) > 0:
        latest_run = data["workflow_runs"][0]
        return {
            "source": "workflow_run",
            "conclusion": latest_run.get("conclusion"),
            "created_at": latest_run.get("created_at"),
            "updated_at": latest_run.get("updated_at"),
            "html_url": latest_run.get("html_url")
        }
    return None

def check_local_manifest(sha):
    manifest_path = os.path.join("StrykerOutput", "stryker-release-manifest.json")
    if os.path.exists(manifest_path):
        try:
            with open(manifest_path, "r", encoding="utf-8") as f:
                manifest = json.load(f)
            if not sha or manifest.get("commit_sha") == sha or manifest.get("commit_sha") == "unknown":
                return {
                    "source": "local_manifest",
                    "passed": manifest.get("passed_gate", False),
                    "score": manifest.get("overall_mutation_score", 0),
                    "min_score": manifest.get("lowest_package_score", 0),
                    "date": manifest.get("execution_date", "N/A"),
                    "status": manifest.get("overall_status", "N/A"),
                    "manifest": manifest
                }
        except Exception as e:
            print(f"Warning: Could not read local manifest: {e}")
    return None

def main():
    target_sha = sys.argv[1] if len(sys.argv) > 1 else os.environ.get("GITHUB_SHA", "")
    config_file = sys.argv[2] if len(sys.argv) > 2 else "stryker-config.json"
    thresholds = load_thresholds(config_file)
    t_break = thresholds["break"]

    repo = os.environ.get("GITHUB_REPOSITORY", "")
    token = os.environ.get("GITHUB_TOKEN", "")

    print("==================================================")
    print("  RELEASE MUTATION TESTING QUALITY GATE AUDITOR   ")
    print(f"  Target Commit SHA: {target_sha or 'HEAD'}")
    print(f"  Break Threshold:   ≥{t_break}%")
    print("==================================================\n")

    commit_analyzed = target_sha or "Unknown"
    execution_date = "N/A"
    mutation_score_text = "N/A"
    passed_break = False
    release_permitted = False
    status_badge = "❌ FAILED"
    verification_details = ""

    local_manifest = check_local_manifest(target_sha)
    if local_manifest:
        commit_analyzed = local_manifest["manifest"].get("commit_sha") or target_sha
        execution_date = local_manifest["date"]
        mutation_score_text = f"{local_manifest['score']:.2f}% (Lowest Package: {local_manifest['min_score']:.2f}%)"
        passed_break = local_manifest["passed"]
        release_permitted = passed_break
        status_badge = local_manifest["status"]
        verification_details = f"Verified from local release manifest ({local_manifest['manifest'].get('packages_count', 0)} packages evaluated)."

    if not release_permitted and repo and target_sha and token and "/" in repo:
        owner, repo_name = repo.split("/", 1)
        commit_status = query_commit_status(owner, repo_name, target_sha, token)
        if commit_status:
            commit_analyzed = target_sha
            execution_date = commit_status.get("updated_at") or commit_status.get("created_at") or "N/A"
            mutation_score_text = commit_status.get("description") or "Recorded in commit status"
            passed_break = (commit_status.get("state") == "success")
            release_permitted = passed_break
            status_badge = "✅ PASSED" if passed_break else "❌ FAILED"
            verification_details = f"Verified from GitHub Commit Status [{commit_status.get('state')}]: {commit_status.get('description')}"
        else:
            workflow_run = query_workflow_runs(owner, repo_name, target_sha, token)
            if workflow_run:
                commit_analyzed = target_sha
                execution_date = workflow_run.get("updated_at") or workflow_run.get("created_at") or "N/A"
                mutation_score_text = f"Workflow Run conclusion: {workflow_run.get('conclusion')}"
                passed_break = (workflow_run.get("conclusion") == "success")
                release_permitted = passed_break
                status_badge = "✅ PASSED" if passed_break else "❌ FAILED"
                verification_details = f"Verified from GitHub Actions workflow run: {workflow_run.get('html_url')}"

    release_decision_text = "✅ RELEASE PERMITTED" if release_permitted else "❌ RELEASE BLOCKED"

    print("[Q1] Which commit was analyzed?")
    print(f"     → {commit_analyzed}")
    print("[Q2] When?")
    print(f"     → {execution_date}")
    print("[Q3] What mutation score was obtained?")
    print(f"     → {mutation_score_text}")
    print(f"[Q4] Did it pass the break threshold (≥{t_break}%)?")
    print(f"     → {'YES' if passed_break else 'NO'}")
    print("[Q5] Can the release proceed?")
    print(f"     → {release_decision_text}\n")

    step_summary_path = os.environ.get("GITHUB_STEP_SUMMARY")
    if step_summary_path:
        summary_md = f"""
# 🚀 Release Mutation Quality Gate Verification

| Release Audit Question | Result |
|------------------------|--------|
| **1. Commit Analyzed** | `{commit_analyzed[:10]}` |
| **2. Execution Date** | {execution_date} |
| **3. Mutation Score Result** | **{mutation_score_text}** |
| **4. Passed Break Threshold (≥{t_break}%)** | {'✅ YES' if passed_break else '❌ NO'} |
| **5. Release Gate Decision** | **{release_decision_text}** |

> **Audit Details**: {verification_details or 'No valid mutation quality gate evidence found for target commit.'}
"""
        with open(step_summary_path, "a", encoding="utf-8") as f:
            f.write(summary_md + "\n")

    github_output_path = os.environ.get("GITHUB_OUTPUT")
    if github_output_path:
        with open(github_output_path, "a", encoding="utf-8") as f:
            f.write(f"release_permitted={'true' if release_permitted else 'false'}\n")
            f.write(f"commit_analyzed={commit_analyzed}\n")
            f.write(f"passed_break={'true' if passed_break else 'false'}\n")

    if not release_permitted:
        print(f"\n❌ RELEASE BLOCKED: Commit {commit_analyzed} has not satisfied the mutation testing quality gate (score >= {t_break}%).", file=sys.stderr)
        print("Please ensure that mutation testing on 'main' has completed successfully before publishing a release.", file=sys.stderr)
        sys.exit(1)

    print(f"✅ RELEASE PERMITTED: Mutation score meets or exceeds break threshold ({t_break}%). Proceeding to package publication.\n")

if __name__ == "__main__":
    main()

File: scripts/verify-stryker-gate.js
// Copyright © Erickson Lopez. MIT License.
const fs = require('fs');
const path = require('path');
const https = require('https');

function loadThresholds(configPath = 'stryker-config.json') {
  let thresholds = { high: 100, low: 98, break: 95 };
  try {
    if (fs.existsSync(configPath)) {
      const config = JSON.parse(fs.readFileSync(configPath, 'utf8'));
      const t = config['stryker-config']?.thresholds || config.thresholds || {};
      thresholds = { high: t.high ?? 100, low: t.low ?? 98, break: t.break ?? 95 };
    }
  } catch (err) {
    console.warn(`Could not parse ${configPath}: ${err.message}`);
  }
  return thresholds;
}

function fetchJson(url, token) {
  return new Promise((resolve, reject) => {
    const urlObj = new URL(url);
    const options = {
      hostname: urlObj.hostname,
      port: 443,
      path: urlObj.pathname + urlObj.search,
      method: 'GET',
      headers: {
        'User-Agent': 'dotnet-idempotency-release-gate',
        'Accept': 'application/vnd.github.v3+json'
      }
    };
    if (token) {
      options.headers['Authorization'] = `Bearer ${token}`;
    }

    const req = https.request(options, (res) => {
      let body = '';
      res.on('data', chunk => { body += chunk; });
      res.on('end', () => {
        if (res.statusCode >= 200 && res.statusCode < 300) {
          try {
            resolve(JSON.parse(body));
          } catch (e) {
            reject(new Error(`Failed to parse JSON response: ${e.message}`));
          }
        } else {
          reject(new Error(`GitHub API request failed with status ${res.statusCode}: ${body}`));
        }
      });
    });

    req.on('error', reject);
    req.end();
  });
}

async function queryCommitStatus(owner, repo, sha, token) {
  const url = `https://api.github.com/repos/${owner}/${repo}/commits/${sha}/statuses`;
  try {
    const statuses = await fetchJson(url, token);
    if (Array.isArray(statuses)) {
      const mutationStatus = statuses.find(s => s.context === 'quality-gate/stryker-mutation');
      if (mutationStatus) {
        return {
          source: 'commit_status',
          state: mutationStatus.state,
          description: mutationStatus.description,
          target_url: mutationStatus.target_url,
          created_at: mutationStatus.created_at,
          updated_at: mutationStatus.updated_at
        };
      }
    }
  } catch (err) {
    console.warn(`Could not query commit statuses: ${err.message}`);
  }
  return null;
}

async function queryWorkflowRuns(owner, repo, sha, token) {
  const url = `https://api.github.com/repos/${owner}/${repo}/actions/workflows/mutation-testing.yml/runs?head_sha=${sha}&status=completed`;
  try {
    const data = await fetchJson(url, token);
    if (data && data.workflow_runs && data.workflow_runs.length > 0) {
      const latestRun = data.workflow_runs[0];
      return {
        source: 'workflow_run',
        conclusion: latestRun.conclusion,
        created_at: latestRun.created_at,
        updated_at: latestRun.updated_at,
        html_url: latestRun.html_url
      };
    }
  } catch (err) {
    console.warn(`Could not query workflow runs: ${err.message}`);
  }
  return null;
}

function checkLocalManifest(sha) {
  const manifestPath = path.join('StrykerOutput', 'stryker-release-manifest.json');
  if (fs.existsSync(manifestPath)) {
    try {
      const manifest = JSON.parse(fs.readFileSync(manifestPath, 'utf8'));
      if (!sha || manifest.commit_sha === sha || manifest.commit_sha === 'unknown') {
        return {
          source: 'local_manifest',
          passed: manifest.passed_gate,
          score: manifest.overall_mutation_score,
          minScore: manifest.lowest_package_score,
          date: manifest.execution_date,
          status: manifest.overall_status,
          manifest
        };
      }
    } catch (err) {
      console.warn(`Could not read local manifest: ${err.message}`);
    }
  }
  return null;
}

async function main() {
  const targetSha = process.argv[2] || process.env.GITHUB_SHA || '';
  const configFile = process.argv[3] || 'stryker-config.json';
  const thresholds = loadThresholds(configFile);

  const repo = process.env.GITHUB_REPOSITORY || '';
  const token = process.env.GITHUB_TOKEN || '';

  console.log(`==================================================`);
  console.log(`  RELEASE MUTATION TESTING QUALITY GATE AUDITOR   `);
  console.log(`  Target Commit SHA: ${targetSha || 'HEAD'}`);
  console.log(`  Break Threshold:   ≥${thresholds.break}%`);
  console.log(`==================================================\n`);

  let commitAnalyzed = targetSha || 'Unknown';
  let executionDate = 'N/A';
  let mutationScoreText = 'N/A';
  let passedBreak = false;
  let releasePermitted = false;
  let statusBadge = '❌ FAILED';
  let verificationDetails = '';

  // 1. Check local manifest first (if available in artifacts)
  const localManifest = checkLocalManifest(targetSha);
  if (localManifest) {
    commitAnalyzed = localManifest.manifest.commit_sha || targetSha;
    executionDate = localManifest.date;
    mutationScoreText = `${localManifest.score}% (Lowest Package: ${localManifest.minScore}%)`;
    passedBreak = localManifest.passed;
    releasePermitted = passedBreak;
    statusBadge = localManifest.status;
    verificationDetails = `Verified from local release manifest (${localManifest.manifest.packages_count} packages evaluated).`;
  }

  // 2. Query GitHub Commit Status API if not already verified or to corroborate
  if (!releasePermitted && repo && targetSha && token) {
    const [owner, repoName] = repo.split('/');
    const commitStatus = await queryCommitStatus(owner, repoName, targetSha, token);
    if (commitStatus) {
      commitAnalyzed = targetSha;
      executionDate = commitStatus.updated_at || commitStatus.created_at;
      mutationScoreText = commitStatus.description || 'Recorded in commit status';
      passedBreak = (commitStatus.state === 'success');
      releasePermitted = passedBreak;
      statusBadge = passedBreak ? '✅ PASSED' : '❌ FAILED';
      verificationDetails = `Verified from GitHub Commit Status [${commitStatus.state}]: ${commitStatus.description}`;
    } else {
      // 3. Check workflow runs API as fallback
      const workflowRun = await queryWorkflowRuns(owner, repoName, targetSha, token);
      if (workflowRun) {
        commitAnalyzed = targetSha;
        executionDate = workflowRun.updated_at || workflowRun.created_at;
        mutationScoreText = `Workflow Run conclusion: ${workflowRun.conclusion}`;
        passedBreak = (workflowRun.conclusion === 'success');
        releasePermitted = passedBreak;
        statusBadge = passedBreak ? '✅ PASSED' : '❌ FAILED';
        verificationDetails = `Verified from GitHub Actions workflow run: ${workflowRun.html_url}`;
      }
    }
  }

  const releaseDecisionText = releasePermitted ? '✅ RELEASE PERMITTED' : '❌ RELEASE BLOCKED';

  console.log(`[Q1] Which commit was analyzed?`);
  console.log(`     → ${commitAnalyzed}`);
  console.log(`[Q2] When?`);
  console.log(`     → ${executionDate}`);
  console.log(`[Q3] What mutation score was obtained?`);
  console.log(`     → ${mutationScoreText}`);
  console.log(`[Q4] Did it pass the break threshold (≥${thresholds.break}%)?`);
  console.log(`     → ${passedBreak ? 'YES' : 'NO'}`);
  console.log(`[Q5] Can the release proceed?`);
  console.log(`     → ${releaseDecisionText}\n`);

  // Write GitHub Step Summary
  const stepSummaryPath = process.env.GITHUB_STEP_SUMMARY;
  if (stepSummaryPath) {
    const summary = `
# 🚀 Release Mutation Quality Gate Verification

| Release Audit Question | Result |
|------------------------|--------|
| **1. Commit Analyzed** | \`${commitAnalyzed.substring(0, 10)}\` |
| **2. Execution Date** | ${executionDate} |
| **3. Mutation Score Result** | **${mutationScoreText}** |
| **4. Passed Break Threshold (≥${thresholds.break}%)** | ${passedBreak ? '✅ YES' : '❌ NO'} |
| **5. Release Gate Decision** | **${releaseDecisionText}** |

> **Audit Details**: ${verificationDetails || 'No valid mutation quality gate evidence found for target commit.'}
`;
    fs.appendFileSync(stepSummaryPath, summary);
  }

  // Set GitHub Outputs
  const outputPath = process.env.GITHUB_OUTPUT;
  if (outputPath) {
    fs.appendFileSync(outputPath, `release_permitted=${releasePermitted}\n`);
    fs.appendFileSync(outputPath, `commit_analyzed=${commitAnalyzed}\n`);
    fs.appendFileSync(outputPath, `passed_break=${passedBreak}\n`);
  }

  if (!releasePermitted) {
    console.error(`\n❌ RELEASE BLOCKED: Commit ${commitAnalyzed} has not satisfied the mutation testing quality gate (score >= ${thresholds.break}%).`);
    console.error(`Please ensure that mutation testing on 'main' has completed successfully before publishing a release.`);
    process.exit(1);
  }

  console.log(`✅ RELEASE PERMITTED: Mutation score meets or exceeds break threshold (${thresholds.break}%). Proceeding to package publication.\n`);
}

main();
